rfcs/20220201-quantization-scaling/README.md

```diff
@@ -0,0 +1,264 @@
+mechanism is aligned with scaling for other post-operations and shares the same
+limitations:
+- only single-value scaling;
+- no run-time support.
+
+API:
+
+```c
+dnnl_status_t DNNL_API dnnl_post_ops_append_binary_v2(dnnl_post_ops_t post_ops,
+        float scale, dnnl_alg_kind_t alg_kind,
+        const dnnl_memory_desc_t *src1_desc);
+
+dnnl_status_t DNNL_API dnnl_post_ops_get_params_binary_v2(
+        const_dnnl_post_ops_t post_ops, int index, float *scale,
+        dnnl_alg_kind_t *alg_kind, const dnnl_memory_desc_t **src1_desc);
+```
+
+An example:
+
+```cpp
+    post_ops p_ops;
+    p_ops.append_binary(scale, algorithm::binary_add, summand_md);
+
+    primitive_attr attr;
+    attr.set_post_ops(p_ops);
+```
+
+#### 2.2.2 Option 2
+
+An alternative solution would be to extend mechanism `attr::scales`. This
+solution doesn't require new API, but extends the existing one.
+
+An example:
+
+```cpp
+    post_ops p_ops;
+    p_ops.append_binary(algorithm::binary_add, summand_md);
+
+    primitive_attr attr;
+    attr.set_post_ops(p_ops);
+    attr.set_scales(DNNL_ARG_ATTR_MULTIPLE_POST_OP(0) | DNNL_ARG_SRC_1,
+            /* mask */ 0, {DNNL_RUNTIME_F32_VAL});
+```
+
+While the alternative solution allows to avoid new API, it introduces one more
+way to set scale which complicates both implementation and integration.
+Please review section
+[Unifying scaling in long-term](#3-unifying-scaling-in-long-term) for more
+details.
+
+
+### 2.3. Scales for inputs in Concat
+
+The mechanism `attr:set_scales` will be introduced to Concat primitive. This
+mechanism is used in Binary primitive, so API is already there.
+
+The following limitations are in place for an initial implementation:
+ - no-runtime support;
+ - mask is 0.
+
+An example:
+
+```cpp
+    primitive_attr attr;
+    for (int n = 0; n < num_src; ++n)
+        attr.set_scales(DNNL_ARG_MULTIPLE_SRC + n, /* mask */ 0, src_scales[n]);
+```
+
+## 3. Unifying scaling in long-term
+
+The motivation behind unifying scaling:
+- Remove duplications;
+- Provide an aligned support across all use cases.
+
+There are a few issues with the current scaling support:
+ - Output scale, scales and binary_mul overlap each other;
+ - RNN quantization is specific to RNN, but can be replaced by other
+ scaling mechanisms:
+   ```
+   # Scaling semantics of set_scales and set_zero_points
+   x_f32 = scale * (x_i8 - zero_point)
+
+   # Scaling semantics of RNN quantization
+   x_i8 = scale * x_f32 + zero_point
+   x_f32 = 1 / scale * (x_i8 - zero_point)
+   ```
+   This will affect application side code as user should pass `1 / scale` to
+   oneDNN comparing to the current `scale`;
+ - Built-in parameter provides very limited flexibility (no mask, no runtime).
+
+The changes:
+1. A unified scaling will be supported using mechanism `attr::set_scales` only:
+ - `attr::output_scales`
+ ``` cpp
+    primitive_attr attr;
+    attr.set_scales(DNNL_ARG_DST, mask, output_scales);
```

Comment:
One more question, how do we define order in which the scales are applied? The `output_scale` is currently applied before post ops.

Reply:
I think what Denis is onto is that we need to clarify when the scales/zero_points are applied. IIUC, the proposal is to do the following:
```
[1] (dst - zp_dst) = post_ops( scale_dst * OP(...) ) 
```
When a user might expect 
```
[2] (dst - zp_dst) * scale_dst = post_ops(OP(...))
```
In the latter case, Denis suggestion (let me know if I misunderstand) is to have a set_output_scales to do the following
```
[3] (dst - zp_dst) * scale_dst = post_ops(output_scales * OP(...))
```

Now some remarks:
- for equation [1], basically we will have to properly document that, as it is not really intuitive (for scale, dst is the output of the OP, whereas for zero_point it is the output of postops)
- for equation [2], if the user wants to rescale before the post-ops, they will have to user binary post-op to rescale after OP
- for equation [3], scale_dst is redundant with the scales of the last post-op when post-ops are present

In my opinion, I think we can go with [2] as it seems to be the cleanest: scales/zero_points are applied to their respective argument, and if a users needs rescaling/reshifting in the post chain, they can rely on binary post-op.

Reply:
Mourad clarified some points offline.

Having argument tag for output scales can be useful if there is a primitive that has multiple output that should be scaled. Currently, we don't have such a primitive but it's possible that the one can be introduced in the future. However, post ops will be affected too as users cannot specify which output a post op can be applied to.

My general concern was that scales for input arguments are clearly and intuitively defined while definition for output scales is counter intuitive. The equation [2] looks reasonable, however, what prevent users from using a post op for scaling the output? In that case the users can define the order in which they scale the output of an operation and we would only need to provide `set_scales` for input arguments in some primitives.

Reply:
+1 for equation 2, which represents the clear quantization definition. Output scales can be the scales of operator output (with w/ or w/o post-ops).

Reply:
Another perspective: without having quantization in mind option 1 makes more sense:

- Option 2 implies solving an equation like `scale * dst = ...`. Why to make it different from source scaling? Why do we have to make it through an equation but not via just direct scaling?
- Putting scaling between applying zero-points and post-ops has to be also carefully documented. Why this order? Why zero-points are so special? Without more context it's similarly not intuitive as in option 2. Again, if we consider it as parts of a bigger quantization scheme, things become more clear.

Reply:
* If we follow option 2 we align scales and zero points (zps are substracted from inputs, but added to outputs)
* zero-point is applied after post-ops, because result of post ops is shifted (for example, relu moves it to positive side), to shift it to s8 we need to shift it back. This is a result of our quantization scheme (`f32 = (s8 - zp_s32) * scale_f32`). In this RFC we generalize quantization, so user on their side can quantize data then pass scale and zero point and expect the right answer, it requires us to follow quantization scheme and do division/addition to destination. As Mourad said, if user is experienced they could use `binary_mul` to avoid division.